auth/service.cc

```diff
@@ -276,34 +276,21 @@ void service::reset_authorization_cache() {
 future<permission_set>
 service::get_uncached_permissions(const role_or_anonymous& maybe_role, const resource& r) const {
     if (is_anonymous(maybe_role)) {
-        return _authorizer->authorize(maybe_role, r);
+        co_return co_await _authorizer->authorize(maybe_role, r);
     }
-
     const std::string_view role_name = *maybe_role.name;
-
-    return has_superuser(role_name).then([this, role_name, &r](bool superuser) {
-        if (superuser) {
-            return make_ready_future<permission_set>(r.applicable_permissions());
-        }
-
-        //
-        // Aggregate the permissions from all granted roles.
-        //
-
-        return do_with(permission_set(), [this, role_name, &r](auto& all_perms) {
-            return get_roles(role_name).then([this, &r, &all_perms](role_set all_roles) {
-                return do_with(std::move(all_roles), [this, &r, &all_perms](const auto& all_roles) {
-                    return parallel_for_each(all_roles, [this, &r, &all_perms](std::string_view role_name) {
-                        return _authorizer->authorize(role_name, r).then([&all_perms](permission_set perms) {
-                            all_perms = permission_set::from_mask(all_perms.mask() | perms.mask());
-                        });
-                    });
-                });
-            }).then([&all_perms] {
-                return all_perms;
-            });
-        });
+    auto superuser = co_await has_superuser(role_name);
+    if (superuser) {
+        co_return r.applicable_permissions();
+    }
+    // Aggregate the permissions from all granted roles.
+    permission_set all_perms;
+    auto all_roles = co_await get_roles(role_name);
+    co_await parallel_for_each(all_roles, [this, &r, &all_perms](std::string_view role_name) -> future<> {
```

Comment:
You seem to be using the regular `parallel_for_each` here. I looked at its implementation and it doesn't seem to keep the lambda alive until all futures spawned by it are resolved. I am afraid you might be running into the lambda coroutine fiasco here because you use the captured `all_perms` which is accessed after a potential preemption.

There exists `coroutine::parallel_for_each` which is reportedly more efficient but can only be waited on in coroutines. It also keeps the lambda alive properly.

Reply:
fixed